src/hooks/useOnClickOutside.ts

```diff
@@ -0,0 +1,21 @@
+import { useEffect } from "react";
+
+export const useOnClickOutside = (
+  ref: React.RefObject<HTMLElement>,
+  cb: () => void
+) => {
+  useEffect(() => {
+    // Bind the event listener
+    document.addEventListener("mousedown", handleClickOutside);
+    return () => {
+      // Unbind the event listener on clean up
+      document.removeEventListener("mousedown", handleClickOutside);
+    };
+  });
```

Comment:
this `useEffect` has no dependencies listed, which means this event binding will occur on every rerender of the component in which it is called...sounds like potential for a memory leak. passing an empty array as dependency will cause event listener to be attached only on mount.

Reply:
> in the screenshot you provide, shapes is a nested group. not all of shapes's children are selected but its box is selected. this would indicate that all children are selected. the parent checkbox should only be checked if all its children are selected.
> 
> i confirmed this UX by using the ant tree select component [here](https://ant.design/components/tree-select/).

The component controlling value is meant to take care of this.